share/shwap/p2p/shrex/server.go

```diff
@@ -0,0 +1,216 @@
+package shrex
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"go.uber.org/zap"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/libs/utils"
+	shrexpb "github.com/celestiaorg/celestia-node/share/shwap/p2p/shrex/pb"
+	"github.com/celestiaorg/celestia-node/store"
+)
+
+// Server implements Server side of shrex protocol to serve data to remote
+// peers.
+type Server struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	host host.Host
+	// list of protocol names that is supported by the server
+	protocols []protocol.ID
+
+	store *store.Store
+
+	params *ServerParams
+	// TODO: decouple middleware metrics from shrex and remove middleware from Server
+	middleware *Middleware
+	metrics    *Metrics
+}
+
+// NewServer creates a new shrEx-Server. It configures the server with the provided
+// parameters, host, and data store. By default, it creates handlers for all types
+// of the requests that the node supports unless the user specifies what protocols should be enabled.
+func NewServer(
+	params *ServerParams,
+	host host.Host,
+	store *store.Store,
+) (*Server, error) {
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("shrex/server: parameters are not valid: %w", err)
+	}
+
+	middleware, err := NewMiddleware(params.ConcurrencyLimit)
+	if err != nil {
+		return nil, fmt.Errorf("shrex/server: could not initialize middleware: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	srv := &Server{
+		ctx:        ctx,
+		cancel:     cancel,
+		store:      store,
+		host:       host,
+		params:     params,
+		middleware: middleware,
+	}
+
+	protocols := make([]protocol.ID, len(registry))
+	index := 0
+	for protocolName, id := range registry {
+		handler := srv.streamHandler(srv.ctx, id)
+		withRateLimit := srv.middleware.rateLimitHandler(ctx, handler, id().Name())
+		withRecovery := RecoveryMiddleware(withRateLimit)
+
+		protocols[index] = ProtocolID(srv.params.NetworkID(), protocolName)
+		srv.SetHandler(protocols[index], withRecovery)
+		index++
+	}
```

Comment:
Set of protocols is not configurable in server constructor right now. Are there plans to make it configurable? If not, registry is not needed and can loop over hardcoded set

Also what is the purpose if  `protocols` array and index var? It sets value to only pass it to SetHandler right away. Looks like some refactoring artifact that is no longer needed.